src/cogent3/core/new_alignment.py

```diff
@@ -0,0 +1,449 @@
+                start=self.start,
+                stop=self.stop,
+                step=self.step,
+                offset=self.offset,
+                seqid=self.seqid,
+                seq_len=self.seq_len,
+            )
+        return self.from_rich_dict(self.to_rich_dict())
+
+
+@dataclass
+class SeqData:
+    __slots__ = ("_data", "_alphabet", "_name_order", "_seq_maker")
+
+    def __init__(
+        self,
+        data: dict[str, SeqTypes],
+        alphabet: CharAlphabet,
+        name_order: Optional[Union[tuple[str], None]] = None,
+        seq_maker: Optional[type] = None,
+    ):
+        self._alphabet = alphabet
+        self._name_order = process_name_order(data, name_order)
+        self._seq_maker = seq_maker
+        # Convert to moltype alphabet indicies
+        self._data: dict[str, numpy.ndarray] = {
+            k: seq_index(v, self._alphabet) for k, v in data.items()
+        }
+
+    @property
+    def set_seq_maker(self) -> type:
+        return self._seq_maker
+
+    @set_seq_maker.setter
+    def set_seq_maker(self, value: tuple[type, CharAlphabet]) -> type:
+        seq_maker, alphabet = value
+        if alphabet != self._alphabet:
+            raise ValueError("Alphabets must be the same")
+        assert alphabet == self._alphabet
+        self._seq_maker = seq_maker
+
+    def _make_seq(self, seqid) -> Union[SeqDataView, type]:
+        seqdataview = self.get_seq_view(seqid=seqid)
+        if self._seq_maker is None:
+            return seqdataview
+        return self.set_seq_maker(seqdataview, seqid=seqid)
+
+    @singledispatchmethod
+    def __getitem__(self, value: Union[str, int]) -> SeqDataView:
+        raise NotImplementedError(f"__getitem__ not implemented for {type(value)}")
+
+    @__getitem__.register
+    def _(self, value: str) -> SeqDataView:
+        return self._make_seq(seqid=value)
+
+    @__getitem__.register
+    def _(self, value: int) -> SeqDataView:
+        seqid = self._name_order[value]
+        return self._make_seq(seqid=seqid)
+
+    def get_seq_array(
+        self, *, seqid: str, start: int = None, stop: int = None
+    ) -> numpy.ndarray:
+        return self._data[seqid][start:stop]
+
+    def get_seq_str(self, *, seqid: str, start: int = None, stop: int = None) -> str:
+        return self._alphabet.from_indices(
+            self.get_seq_array(seqid=seqid, start=start, stop=stop)
+        )
+
+    def get_seq_bytes(
+        self, *, seqid: str, start: int = None, stop: int = None
+    ) -> bytes:
+        return self.get_seq_str(seqid=seqid, start=start, stop=stop).encode("utf8")
+
+    def iter_names(self, *, name_order: tuple[str] = None) -> Iterator:
+        yield from process_name_order(self._name_order, name_order)
+
+    def get_seq_view(self, seqid: str) -> SeqDataView:
+        seq_len = len(self._data[seqid])
+        return SeqDataView(seq=self, seqid=seqid, seq_len=seq_len)
+
+    def iter_seq_view(self, *, name_order: tuple[str] = None) -> Iterator:
```

Comment:
Delete this method for now. (I suspect we may get it "for free" having implemented `__getitem__`)